server/src/main/java/org/elasticsearch/index/shard/IndexLongFieldRange.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+import static org.elasticsearch.index.shard.ShardLongFieldRange.LONG_FIELD_RANGE_VERSION_INTRODUCED;
+
+/**
+ * Class representing an (inclusive) range of {@code long} values in a field in an index which may comprise multiple shards. This
+ * information is accumulated shard-by-shard, and we keep track of which shards are represented in this value. Only once all shards are
+ * represented should this information be considered accurate for the index.
+ */
+public class IndexLongFieldRange implements Writeable, ToXContentFragment {
+
+    /**
+     * Sentinel value indicating that no information is currently available, for instance because the index has just been created.
+     */
+    public static final IndexLongFieldRange UNKNOWN = new IndexLongFieldRange(new int[0], Long.MAX_VALUE, Long.MIN_VALUE);
+
+    /**
+     * Sentinel value indicating an empty range, for instance because the field is missing or has no values in any shard.
+     */
+    public static final IndexLongFieldRange EMPTY = new IndexLongFieldRange(null, Long.MAX_VALUE, Long.MIN_VALUE);
+
+    /**
+     * Sentinel value indicating the actual range may change in the future.
+     */
+    public static final IndexLongFieldRange MUTABLE = new IndexLongFieldRange(null, Long.MIN_VALUE, Long.MAX_VALUE);
```

Comment:
I would prefer to rename MUTABLE to UNKNOWN and use UNFILLED (or a nicer english word) for the no information case.

I think UNKNOWN better captures the state where we do started shards but do not know the range.

Reply:
I initially rejected `UNKNOWN` here since I felt it might be taken as an implication that it might become known in future, whereas `MUTABLE` is more immutable :grin: But I'm ok with this renaming. I've renamed `UNKNOWN` to `NO_SHARDS` too.